Enrich test results produced by test.py

xtrey opened the pull request:
Related to #17851

This PR resolves issue with double count of the test result for topology tests. It will not appear in the consolidated report anymore.
Another fix is to provide a better view which test failed by modifying the test case name in the report enriching it with mode and run id, so making them unique across the run.

The scope of this change is:
1. Modify the test name to have run id in name
2. Add handlers to get logs of test.py and pytest in one file that are related to test, rather than to the full suite
3. Remove topology tests from aggregating them on a suite level in Junit results
4. Add a link to the logs related to the failed tests in Junit results, so it will be easier to navigate to all logs related to test
5. Gather logs related to the failed test to one directory for better logs investigation

temichus commented on test/pylib/log_browsing.py:
```diff
@@ -100,3 +100,13 @@ async def grep(self, expr: str | re.Pattern, filter_expr: Optional[str | re.Patt
                 line = await self._run_in_executor(f.readline, loop=loop)
         return matchings
+
+    async def read_lines(self, from_mark=None, to_mark=None):
```
add docstring for this

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8027/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8027/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8027/)
#### Build Details:
- Duration: 4 hr 32 min
- Builder: i-06317f8bad34c0b84 (m5ad.12xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8037/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8037/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8037/)
#### Build Details:
- Duration: 2 hr 30 min
- Builder: spider8.cloudius-systems.com

kbr-scylla commented on test.py:
```diff
@@ -1019,9 +1022,13 @@ def __init__(self, test_no: int, shortname: str, casename: str, suite) -> None:
     async def run(self, options: argparse.Namespace) -> Test:
 
         self._prepare_pytest_params(options)
+        architecture = self.suite.options.tmpdir.split(os.sep)[-1] if self.suite.options.tmpdir.split(os.sep)[-1] in ["x86_64", "aarch64"] else None
```
We should not reverse dependencies like that.
Architecture of the solution should be layered and dependencies should go in one direction.

test.py doesn't know whether its being run from jenkins or locally by a developer or from some other application or whatever. The implementation of test.py must not depend on such things.

xtrey replied:
As for now, test.py does not know such things, but it should if we want to add some features. Since the architecture is a part of a temporary directory parameter that CI passes to the test.py we need to know to be able to create directories in the correct place. Even for the creation of the directory we do not need this information, but we need to construct the URL that will be attached to the failed report as a link to the directory with all logs. As far as I remember this is what quite a lot of developers wanted.
I'll modify this logic to remove the architecture parameter.

kbr-scylla replied:
This can be done in test.py without introducing knowledge about the CI environment / reversed dependency.

For example, to determine the architecture, instead of parsing tmpdir path -- which depends on the specific fact that CI environment runs tests under a path which contains architecture in its name, and is not true in general (e.g. local developer runs) -- we can introduce a parameter to test.py, e.g. `--arch`, which Jenkins pipeline would then pass to test.py.

See also https://en.wikipedia.org/wiki/Dependency_injection

kbr-scylla replied:
In fact, for the particular case of architecture, we don't need a parameter...
https://docs.python.org/3/library/platform.html

kbr-scylla replied:
> we need to construct the URL that will be attached to the failed report as a link to the directory with all logs

Determining the URL belongs inside Jenkins pipeline logic, I believe.

test.py could then receive the URL as part of its configuration.

xtrey replied:
we are still depending on what directory will be passed from CI and how it will be copied to artifacts later.
you are right that I do not need the architecture. I've modified already the logic of building the URL.

xtrey replied:
we are still depending on what directory will be passed from CI and how it will be copied to artifacts later.
you are right that I do not need the architecture. I've modified already the logic of building the URL.

> URL

This is what I'm doing. I'm getting the BUILD_URL from environment variables that will give us the URL up to the build number and then construct the rest of the path.

kbr-scylla replied:
> This is what I'm doing. I'm getting the BUILD_URL from environment variables that will give us the URL up to the build number and then construct the rest of the path.

What about this:
```
        return f"<a href={build_url}artifact/{dir_path_relative}</a>"
```

Isn't the "artifact" directory name specific to Jenkins?

I'm not sure I understand why this "logs dir link" needs to be calculated inside test.py. Can't it be passed from Jenkins pipeline as parameter?


Also, why are you using environment variable, when most of test.py configuration is passed through commandline parameters or yaml configuration? We should minimize the number of different ways to pass parameters.

xtrey replied:
> > This is what I'm doing. I'm getting the BUILD_URL from environment variables that will give us the URL up to the build number and then construct the rest of the path.
> 
> What about this:
> 
> ```
>         return f"<a href={build_url}artifact/{dir_path_relative}</a>"
> ```
> 
> Isn't the "artifact" directory name specific to Jenkins?
> 
> I'm not sure I understand why this "logs dir link" needs to be calculated inside test.py. Can't it be passed from Jenkins pipeline as parameter?
> 
> Also, why are you using environment variable, when most of test.py configuration is passed through commandline parameters or yaml configuration? We should minimize the number of different ways to pass parameters.

you are right, the artifacts directory is Jenkins-specific. but this property anyway needed only on the Jenkins report.
it should be done inside test.py because test.py is aware of the fact which test failed and what it should contain in that directory. 
It can be passed in jenkins pipeline. But why do we need an additional argument when we can extract it directly from the environment variable? Especially since it's a common way to get a build URL in Jenkins

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -127,13 +131,42 @@ async def before_test(self, test_case_name: str) -> None:
 
+    async def mark_scylla_log(self, test_case_name: str, test_log: Path) -> None:
+        """
+        Mark the place in the log file to be able to extract only part related to the test and add a handler for test
```
I disagree with the idea of extracting only part of the log.

A failure could be related to something that happened before the test started. For example, a bug during node bootstrap, which happened during cluster setup phase, before test started using the cluster.

xtrey replied:
We are attaching it before the test phase so it will be catched there. In the worst-case scenario, we have in the directory the full Scylla log as well.

kbr-scylla commented on test/topology/conftest.py:
```diff
@@ -168,18 +171,43 @@ async def manager_internal(event_loop, request):
 
+
+def build_logs_dir_link(failed_test_dir_path: pathlib.Path):
+    """Build the link to the logs directory in artifacts for the failed test in a Jenkins report"""
```
This code does not belong inside test.py.
And it doesn't apply when we run outside Jenkins.

test.py should not even know what "Jenkins" is

xtrey replied:
How we can add Jenkins-specific features without knowing that we are on Jenkins? :thinking: 
This code will be executed regardless of where it was started: dev's machine or Jenkins. The only functionality that differs is URL generation which will be set to N/A in case when run was on the dev's machine.

kbr-scylla replied:
> How we can add Jenkins-specific features without knowing that we are on Jenkins? 🤔

The code that is specific to Jenkins pipeline implementation belongs inside the Jenkins pipeline implementation.

If the pipeline needs additional parameters to pass to test.py, you can add those parameters to test.py. But test.py doesn't know who is calling it and who is passing the parameters. It could be a developer running locally or it could be automation, Jenkins or something else.

Suppose that in the future we want to modify jenkins pipeline to change how the paths are named. If test.py contains code that depends on jenkins pipeline implementation details, such as path naming, the test.py code will break. On the other hand, if it provides configuration options that jenkins pipeline uses, then you will simply modify the values of those parameters in the same jenkins pipeline commit where you change path names.

Or, suppose that in the future we'd like to execute test.py from within GitHub actions instead of Jenkins. Instead of adding yet more code to test.py to support GitHub actions, you don't need to do anything, because test.py is already providing necessary configuration options, you can just provide appropriate values when calling it from GitHub actions.

It's a software engineering principle to write code in modularized and layered way. If layer X depends on layer Y, then layer Y does not know about the implementation of layer X. In fact layer X doesn't know about implementation of Y either, it only knows the interface of Y.

kbr-scylla commented on test/pylib/log_browsing.py:
```diff
@@ -100,3 +100,17 @@ async def grep(self, expr: str | re.Pattern, filter_expr: Optional[str | re.Patt
+        """
+        Returns a list of lines from the Scylla log file that are between provided marks.
+        """
+        lines: list[str] = []
```
The log file may be large, I'm afraid in some cases it may not fit into memory

kbr-scylla replied:
The function could be easily modified to return a generator of lines, which you could then use at call site to write line by line

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8049/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8049/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: object_store
:small_blue_diamond: pylib
:small_blue_diamond: pylib
:small_blue_diamond: pylib
:small_blue_diamond: topology
:small_blue_diamond: topology_custom/test_multidc
#### Failed Tests (4/10220):
- [test_remove_node_add_column.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8049/testReport/junit/%28root%29/test_topology_remove_decom/test_remove_node_add_column_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_node_add_column.release.2)
- [test_remove_node_add_column.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8049/testReport/junit/%28root%29/test_topology_remove_decom/test_remove_node_add_column_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_node_add_column.release.3)
- [test_remove_node_add_column.release.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8049/testReport/junit/%28root%29/test_topology_remove_decom/test_remove_node_add_column_release_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_node_add_column.release.4)
- [test_raft_recovery_basic.debug.22](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8049/testReport/junit/%28root%29/test_raft_recovery_basic/test_raft_recovery_basic_debug_22) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_recovery_basic.debug.22)
#### Build Details:
- Duration: 10 hr
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8075/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8075/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: pylib
:small_blue_diamond: topology_custom/test_multidc
#### Failed Tests (300/1800):
- [test_query_dc_with_rf_0_does_not_crash_db.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8075/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db.debug.1)
- [test_query_dc_with_rf_0_does_not_crash_db.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8075/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db.debug.10)
- [test_query_dc_with_rf_0_does_not_crash_db.debug.100](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8075/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db_debug_100) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db.debug.100)
- [test_query_dc_with_rf_0_does_not_crash_db.debug.11](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8075/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db_debug_11) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db.debug.11)
- [test_query_dc_with_rf_0_does_not_crash_db.debug.12](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8075/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db_debug_12) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db.debug.12)
- [test_query_dc_with_rf_0_does_not_crash_db.debug.13](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8075/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db_debug_13) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db.debug.13)
- [test_query_dc_with_rf_0_does_not_crash_db.debug.14](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8075/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db_debug_14) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db.debug.14)
- [test_query_dc_with_rf_0_does_not_crash_db.debug.15](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8075/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db_debug_15) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db.debug.15)
- [test_query_dc_with_rf_0_does_not_crash_db.debug.16](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8075/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db_debug_16) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db.debug.16)
- [test_query_dc_with_rf_0_does_not_crash_db.debug.17](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8075/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db_debug_17) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db.debug.17)
#### Build Details:
- Duration: 1 hr 21 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8196/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8196/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: pylib
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8196/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8196/)
#### Build Details:
- Duration: 2 hr 11 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8211/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8211/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: pylib
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8211/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8211/)
#### Build Details:
- Duration: 3 hr 19 min
- Builder: i-023213232eda0e596 (m5d.12xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8306/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8306/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_3_x_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8306/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8306/)
#### Build Details:
- Duration: 2 hr 31 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8340/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8340/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8340/)
#### Build Details:
- Duration: 2 hr 22 min
- Builder: spider7.cloudius-systems.com

xtrey commented:
@kbr-scylla @piodul This is how it will look with current changes and with changes from pipeline merged https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/35/
Please take a look and let me know if there is any information missing/redundant in the provided logs.

mykaul commented:
I'm not sure why TEST_LOGS doesn't end up as a URL, but as HTML, which makes me always need to copy-paste it instead of clicking on it. Perhaps unrelated to this change though.

temichus commented:
> I'm not sure why TEST_LOGS doesn't end up as a URL, but as HTML, which makes me always need to copy-paste it instead of clicking on it. Perhaps unrelated to this change though.

@mykaul,  not sure about what HTML you are talking about. but after this PR we will have a link to failed_test_logs https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/35/testReport/(root)/test_multidc/test_multidc_debug_1/ 
but only for topology tests for now

mykaul commented:
Example:
![image](https://github.com/scylladb/scylladb/assets/4655593/59c7ca3f-e2ab-4bad-8a2a-7f6777558e21)

xtrey commented:
> Example: ![image](https://private-user-images.githubusercontent.com/4655593/326243401-59c7ca3f-e2ab-4bad-8a2a-7f6777558e21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5ODg5NzAsIm5iZiI6MTcxNDk4ODY3MCwicGF0aCI6Ii80NjU1NTkzLzMyNjI0MzQwMS01OWM3Y2EzZi1lMmFiLTRiYWQtOGEyYS03ZjY3Nzc1NThlMjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDZUMDk0NDMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjFlMzU3ZmIzM2YyMDMyYTU3ZTZmYmIzYmYyYTM0MjI0ODYzYjhmMjRlZjgxM2U2ZGE2MzBiYjc2ZWRlNDQwZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.vsrJpMiCqpfbu56idlmmMXy4lVKQ0iJ3or-N3t0b5p8)

that's funny, since if you click the failed test case there will be a normal link and not just an HTML tag. I'll try to play around to make it work in both cases though.

xtrey commented:
@mykaul seems this is the limitation of the plugin in Jenkins. https://jenkins.scylladb.com/job/sct-github-PRs-scan/job/scylla-dtest/job/PR-3831/35/testReport/hintedhandoff_additional_test/ dtest have the same issue with this, so I suppose we will live with that. 
@kbr-scylla @piodul what is your opinion about that?

kbr-scylla commented:
> @kbr-scylla @piodul This is how it will look with current changes and with changes from pipeline merged https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/35/

I see
![Screenshot from 2024-05-07 15-09-59](https://github.com/scylladb/scylladb/assets/120044486/b04dd696-a242-467d-97f6-ec2494c67ba1)

The "failed test logs" link directs to https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/35/artifact/testlog/x86_64/debug/failed_test/test_multidc.debug.1
![Screenshot from 2024-05-07 15-10-32](https://github.com/scylladb/scylladb/assets/120044486/5c840741-fdca-4547-97b6-98c7164dcee9)


Got GCd already?

Also, please try what will happen if there are two test cases in a single file and both fail.

kbr-scylla commented:
Started rebuild hope it will work: https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/38/

xtrey commented:
@kbr-scylla seems it was GCed. your run didn't pass because I've changed the job configuration to check another thing. restarted it with the correct config and added a second fail in the file. https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/39/

kbr-scylla commented:
Looks very good (and a huge improvement over current state I think) but there are some problems:

https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/39/testReport/junit/(root)/test_multidc/test_putget_2dc_with_rf_nodes_list0_1__debug_1/

-> click failed_test_logs, we get
https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/39/artifact/testlog/x86_64/debug/failed_test/test_putget_2dc_with_rf[nodes_list0-1].debug.1/

-> there are two test log files (besides Scylla logs):
1. first file: [topology_custom.test_multidc.1.log](https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/39/artifact/testlog/x86_64/debug/failed_test/test_putget_2dc_with_rf[nodes_list0-1].debug.1/topology_custom.test_multidc.1.log)

this is a part of the pytest log. But it's missing the  FAILURES section where the actual failure is printed, which I believe is important.
You can find the full log file here: https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/39/artifact/testlog/x86_64/debug/topology_custom.test_multidc.1.log and it contains the FAILURES section. Part of it relevant for this test case should be included in our file as well:
```
=================================== FAILURES ===================================
_________________________________ test_multidc _________________________________
... bla bla ...
____________________ test_putget_2dc_with_rf[nodes_list0-1] ____________________

... THIS is the part that should end up in our file for this particular test failure ...
... all the way until ...

____________________ test_putget_2dc_with_rf[nodes_list1-2] ____________________
... bla bla ...
```

2. second file: [topology_custom.test_multidc.1.test_putget_2dc_with_rf[nodes_list0-1].log](https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/39/artifact/testlog/x86_64/debug/failed_test/test_putget_2dc_with_rf[nodes_list0-1].debug.1/topology_custom.test_multidc.1.test_putget_2dc_with_rf%5Bnodes_list0-1%5D.log)

this is a part the test.py log. But it looks like it's both missing some things and having some extra unnecessary things:
- it doesn't seem to have all `[debug/topology_custom.test_multidc.1]` lines?
the final line with this is:
```
23:41:39.148 INFO> [debug/topology_custom.test_multidc.1] Cluster ScyllaCluster(name: 3087d928-0cb2-11ef-9583-a8a159ead531, running: ScyllaServer(1047, 127.67.75.19, cdb1995c-3649-4de8-8e51-927b21f89bdf), ScyllaServer(1048, 127.67.75.3, 3346eed4-9cd9-4332-aa23-626e1679749d), stopped: ) added ScyllaServer(1048, 127.67.75.3, 3346eed4-9cd9-4332-aa23-626e1679749d)
```
But if you look at full log: https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/39/artifact/testlog/x86_64/test.py.debug-release-dev.log

then the final line for cluster used in this test is:
```
23:41:42.650 INFO> [debug/topology_custom.test_multidc.1] Cluster ScyllaCluster(name: 3087d928-0cb2-11ef-9583-a8a159ead531, running: , stopped: ScyllaServer(1047, 127.67.75.19, cdb1995c-3649-4de8-8e51-927b21f89bdf), ScyllaServer(1048, 127.67.75.3, 3346eed4-9cd9-4332-aa23-626e1679749d)) releases ips {'127.67.75.19', '127.67.75.3'}
```
this line is missing from our cut file (and many lines before it are also missing).

I think we should include everything that happens to the cluster that was used in this test, all the way until the cluster released from the test. (It may be later reused in another test case --- we don't need to include that part, I guess)
- it contains some lines related to other test: 
```
23:41:40.074 INFO> [debug/topology_custom.test_tablets_migration.2] _cluster_remove_node ScyllaServer(1035, 127.67.75.23, cf8d8c05-84c0-448f-beb8-531a50780346) with initiator ScyllaServer(1033, 127.67.75.15, a2c7a0d7-ff48-432f-a7a8-ee0328a2db36)
```
- the name of this file, I think should reflect the fact that it comes from the test.py log. Maybe include the string `test.py` in the file name?

BTW it's weird that the name of this file includes the name of the test case `test_putget_2dc_with_rf[nodes_list0-1]` but the other file (pytest) does not. If anything it should be the other way around.

Alternatively, since we already know which test we're dealing with from the directory, maybe we could simply name them `pytest.log` and `test.py.log`.

xtrey commented:
@kbr-scylla 
1. This information missing since it's printed by pytest itself and not the test. This stack trace is printed on the summary page of the test and it's not a log actually, it's additional data of the test rather than a log. However, if you need it there I guess I can put it explicitly in after the test method.
2. That's correct and that's intentional. ```debug/topology_custom.test_multidc.1``` is for the whole file topology_custom/test_multidc that have 4 test cases. So you have only logs related to the specific test case that failed rather than all logs from the whole file.
You can see debug/topology_custom.test_tablets_migration.2 in the test.py log as well. This is where the after-test method finished for the failed test, we have several threads in one pytest instance. 
You are comparing different states of the cluster: one is after one test, and another one is after another test. This is visible from the timestamps.

One log file is pytest as you mentioned, but anyway, it's a part of the test.py log. The limitation is that currently, I'm copying the test.py log when the test method finishes, so not the whole test.py log will be copied. 

> I think we should include everything that happens to the cluster that was used in this test, all the way until the cluster released from the test.
Could you please elaborate on this point what exactly you are lacking? 

The whole idea of this change was to reduce information noise in logs. To achieve that I wanted to gather logs from pytest and Scylla that are **strictly related** to the **specific test method** from the file, and not the whole file with a bunch of the test methods. So do you need logs in one place for the whole file or for each test case individually?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8716/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8716/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_multidc
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8716/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1515/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9537/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8716/)
#### Failed Tests (1/31248):
- [test_tablet_split.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8716/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_split_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_split.debug.4)
#### Build Details:
- Duration: 3 hr 13 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8717/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8717/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_multidc
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8717/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8717/)
#### Build Details:
- Duration: 6 hr 26 min
- Builder: spider5.cloudius-systems.com

xtrey commented:
Changes from the previous commit:
1. Rebase latest master
2. Rename the log files in the failed_directory
3. Add stacktrace
4. Revert several unwanted changes

kbr-scylla commented on test/topology/conftest.py:
```diff
@@ -168,18 +180,44 @@ async def manager_internal(event_loop, request):
+    tmp_dir = pathlib.Path(request.config.getoption('tmpdir'))
+    xml_path = pathlib.Path(request.config.getoption('xmlpath'))
+    suite_testpy_log = tmp_dir / mode / f"{xml_path.name[:-10]}.log"
+    test_log = tmp_dir / mode / f"{suite_testpy_log.name[:-4]}.{test_case_name}.log"
```
Those `-10` and `-4` magic numbers should have a comment.

Better yet if there was another way to calculate it without relying on hardcoded lengths. What is this doing exactly? Are we trying to cut of some particular suffix? Are we cutting of everything after the last dot, or something like that?

kbr-scylla replied:
Or:

https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.stem

kbr-scylla commented on test/topology/conftest.py:
```diff
@@ -168,18 +180,44 @@ async def manager_internal(event_loop, request):
+            f.write(request.node.stash[phase_report_key].longreprtext)
+        if request.config.getoption('artifacts_dir_url'):
+            # get the relative path to the tmpdir for the failed directory
+            dir_path_relative = f"{failed_test_dir_path.as_posix()[failed_test_dir_path.as_posix().find('testlog'):]}"
```
is there no better way to calculate this (using some nice high-level methods from `Path`) instead of these string manipulations?

xtrey replied:
The current way of calculating the path is fragile and clumsy, but I could not find a better solution. An alternative solution was introducing additional parameter to the test.py and providing it during the run, but I'm not a big fan of such an idea.

kbr-scylla replied:
ok, so let's leave it like that

kbr-scylla commented on test/topology/conftest.py:
```diff
@@ -168,18 +180,44 @@ async def manager_internal(event_loop, request):
+            full_url = f"<a href={request.config.getoption('artifacts_dir_url')}/{dir_path_relative}>failed_test_logs</a>"
+            record_property("TEST_LOGS", full_url)
+
+    os.unlink(test_log)
```
There should be a `Path` method for that

kbr-scylla commented on test/topology/conftest.py:
```diff
@@ -226,3 +264,24 @@ def skip_mode_fixture(request, mode):
+        outcome = yield
+        report = outcome.get_result()
+        for param in [self.mode, self.run_id]:
+            if param:
```
Is it possible we don't have `mode` or `run_id`?

temichus replied:
mode is required
```
 parser.addoption('--mode', action='store', required=True,
                     help='Scylla build mode. Tests can use it to adjust their behavior.')
```
run_id is None by default
```
  parser.addoption('--run_id', action='store', default=None,
```
so "if param" will false in the case of Node

kbr-scylla replied:
Agreed with using default -- and then we can remove this weird `for` loop, instead do `report.node_id = f"{report.nodeid}.{self.mode}.{self.run_id or 1}"` or sth

(but probably better to assign `1` in constructor when assigning `self.run_id` and then it doesn't have to be `Optional`)

xtrey commented on test/pylib/manager_client.py:
```diff
@@ -109,8 +112,18 @@ def _driver_update(self) -> None:
+        self.test_log_fh = logging.FileHandler(test_log, mode='w+')
+        # to get the formatter with a timestamp we need to add explicitly here for our handler
+        # to make it consistent with the root logger formatter we extract it and applying to the handler
+        self.test_log_fh.setFormatter(logging.getLogger().handlers[0].formatter)
```
We have formatter set on the root logger to have timestamps and suffix of the test suite, that's why I'm getting this formatter and applying it here. 
```logger.getLogger()``` returns the root logger and while ```test_logger``` is different, since I'm explicitly getting it.

kbr-scylla replied:
Ok. (Could be nice to have this explained in the comment though.)

xtrey replied:
There is already a comment on why we are getting formatter from the root logger. I'll try to rephrase it since it seems not clear.

xtrey commented:
Fix code review comments.
Remove magic log path slicing with Path.stem
Rephrase the comment about why we need formatter from root logger
Remove FAILED_KEY since we using the whole report
Set default value for run_id

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8843/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8843/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8843/)
#### Build Details:
- Duration: 6 hr 16 min
- Builder: i-02e3b17be1ac57a0d (m5ad.8xlarge)

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -109,8 +112,18 @@ def _driver_update(self) -> None:
-    async def before_test(self, test_case_name: str) -> None:
-        """Before a test starts check if cluster needs cycling and update driver connection"""
+    async def before_test(self, test_case_name: str, test_log: Path) -> None:
+        # Pytest uses 'test' handler to for output logs during tests.
```
nit: forgot to update comment

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -129,11 +142,19 @@ async def before_test(self, test_case_name: str) -> None:
 
     async def after_test(self, test_case_name: str, success: bool) -> None:
         """Tell harness this test finished"""
+        logging.getLogger('test').removeHandler(self.test_log_fh)
```
You add the handler to root but remove from `test`, I'm guessing this call has no effect so we're leaking the handler?

kbr-scylla commented:
I like the results now: https://jenkins.scylladb.com/job/scylla-staging/job/xtrey/job/scylladb_test.py/46/

kbr-scylla commented:
Also needs rebase (after @temichus's PR went in)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8933/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8933/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8933/)
#### Build Details:
- Duration: 4 hr 54 min
- Builder: i-0032bb47f97ea984b (m5ad.12xlarge)
